File: sdk/typescript/src/run-helpers.ts
import type { Schema } from './client.js';

export class RunFailedError extends Error {
  constructor(
    readonly runId: string,
    readonly status: string,
    readonly result: Schema['RunResult'],
    readonly failureCode?: string,
  ) {
    super(`Run ${runId} ended with status ${status} (persistence: ${result.persistence_status}).`);
    this.name = 'RunFailedError';
  }
}
export class WaitTimeoutError extends Error {
  constructor(readonly runId: string) {
    super(`Timed out waiting for run ${runId}. The agent has not been cancelled.`);
    this.name = 'WaitTimeoutError';
  }
}
export type WaitOptions = { timeoutMs?: number; pollIntervalMs?: number; signal?: AbortSignal };
type Runs = {
  get(id: string, options?: { signal?: AbortSignal }): Promise<Schema['Run']>;
  getResult(id: string, options?: { signal?: AbortSignal }): Promise<Schema['RunResult']>;
};
const terminal = new Set(['succeeded', 'failed', 'cancelled', 'timed_out']);

export async function waitForRun(runs: Runs, runId: string, options: WaitOptions = {}) {
  const interval = options.pollIntervalMs ?? 1000;
  if (!Number.isFinite(interval) || interval <= 0) throw new RangeError('pollIntervalMs must be positive');
  if (
    options.timeoutMs !== undefined &&
    (!Number.isFinite(options.timeoutMs) || options.timeoutMs < 0 || options.timeoutMs > 2_147_483_647)
  )
    throw new RangeError('timeoutMs must be between 0 and 2147483647');
  const timeout = new AbortController();
  const failure = new WaitTimeoutError(runId);
  const timer =
    options.timeoutMs === undefined ? undefined : setTimeout(() => timeout.abort(failure), options.timeoutMs);
  if (options.timeoutMs === 0) timeout.abort(failure);
  const signal = options.signal ? AbortSignal.any([options.signal, timeout.signal]) : timeout.signal;
  try {
    for (;;) {
      signal.throwIfAborted();
      const run = await runs.get(runId, { signal });
      if (terminal.has(run.status)) {
        const result = await runs.getResult(runId, { signal });
        signal.throwIfAborted();
        if (result.final && result.persistence_status !== 'pending') {
          if (
            run.status !== 'succeeded' ||
            result.execution_outcome !== 'success' ||
            !['verified', 'not_required'].includes(result.persistence_status)
          )
            throw new RunFailedError(runId, run.status, result, run.failure_code);
          return result;
        }
      }
      await new Promise<void>((resolve, reject) => {
        const abort = () => {
          clearTimeout(poll);
          reject(signal.reason);
        };
        const poll = setTimeout(() => {
          signal.removeEventListener('abort', abort);
          resolve();
        }, interval);
        signal.addEventListener('abort', abort, { once: true });
      });
    }
  } finally {
    clearTimeout(timer);
  }
}

export async function* streamRunText(
  events: AsyncGenerator<Schema['Event']>,
  complete: () => Promise<Schema['RunResult']>,
): AsyncGenerator<string> {
  // for-await closes the source on detachment; completion is checked only after normal EOF.
  for await (const event of events) {
    if (event.type === 'output.delta' && typeof event.data.text === 'string' && event.data.text)
      yield event.data.text;
  }
  await complete();
}

File: sdk/python/macrofold/run_helpers.py
"""Presentation and completion helpers over the maintained transport and typed resources."""
import math
import time
from collections.abc import Callable, Generator
from typing import TYPE_CHECKING
from .models import Event, Run, RunResult
from .resource_options import decode

if TYPE_CHECKING:
    from .client import Client


class RunFailedError(Exception):
    def __init__(self, run_id: str, status: str, result: RunResult, failure_code: str | None = None):
        self.run_id, self.status, self.result, self.failure_code = run_id, status, result, failure_code
        super().__init__(f'Run {run_id} ended with status {status} (persistence: {result.persistence_status}).')


class WaitTimeoutError(TimeoutError):
    def __init__(self, run_id: str):
        self.run_id = run_id
        super().__init__(f'Timed out waiting for run {run_id}. The agent has not been cancelled.')


def wait_for_run(client: 'Client', run_id: str, *, timeout: float | None = None, poll_interval: float = 1) -> RunResult:
    if not math.isfinite(poll_interval) or poll_interval <= 0:
        raise ValueError('poll_interval must be positive')
    if timeout is not None and (not math.isfinite(timeout) or timeout < 0):
        raise ValueError('timeout must be a non-negative number')
    deadline = None if timeout is None else time.monotonic() + timeout
    try:
        while True:
            if deadline is not None and time.monotonic() >= deadline:
                raise WaitTimeoutError(run_id)
            run = decode(Run, client.request('getRun', path={'run_id': run_id}, _deadline=deadline), None)
            if run.status in {'succeeded', 'failed', 'cancelled', 'timed_out'}:
                result = decode(RunResult, client.request('getRunResult', path={'run_id': run_id}, _deadline=deadline), None)
                if result.final and result.persistence_status != 'pending':
                    if run.status != 'succeeded' or result.execution_outcome != 'success' or result.persistence_status not in {'verified', 'not_required'}:
                        raise RunFailedError(run_id, run.status, result, run.failure_code)
                    return result
            delay = poll_interval if deadline is None else min(poll_interval, max(0, deadline - time.monotonic()))
            time.sleep(delay)
    except TimeoutError as error:
        if isinstance(error, WaitTimeoutError):
            raise
        raise WaitTimeoutError(run_id) from error


def stream_run_text(events: Generator[Event, None, None], complete: Callable[[], RunResult]) -> Generator[str, None, None]:
    try:
        for event in events:
            text = event.data.get('text')
            if event.type == 'output.delta' and isinstance(text, str) and text:
                yield text
    finally:
        events.close()
    complete()
